lxd/storage_zfs.go

```diff
@@ -542,42 +543,45 @@ func (s *storageZfs) ContainerCreate(container container) error {
 func (s *storageZfs) ContainerCreateFromImage(container container, fingerprint string) error {
 	shared.LogDebugf("Creating ZFS storage volume for container \"%s\" on storage pool \"%s\".", s.volume.Name, s.pool.Name)
 
-	containerPath := container.Path()
-	containerName := container.Name()
-	fs := fmt.Sprintf("containers/%s", containerName)
-	containerPoolVolumeMntPoint := getContainerMountPoint(s.pool.Name, containerName)
+	containerCreateFromImagePoolLockID := getContainerCreateFromImageLockID(s.pool.Name, fingerprint)
 
-	fsImage := fmt.Sprintf("images/%s", fingerprint)
+	lxdContainerCreationLock.Lock()
+	_, ok := lxdContainerCreationMap[containerCreateFromImagePoolLockID]
+	if !ok {
+		lxdContainerCreationMap[containerCreateFromImagePoolLockID] = &containerCreationStruct{}
+		lxdContainerCreationMap[containerCreateFromImagePoolLockID].lock = &sync.RWMutex{}
+	}
 
-	imageStoragePoolLockID := getImageCreateLockID(s.pool.Name, fingerprint)
-	lxdStorageMapLock.Lock()
-	if waitChannel, ok := lxdStorageOngoingOperationMap[imageStoragePoolLockID]; ok {
-		lxdStorageMapLock.Unlock()
-		if _, ok := <-waitChannel; ok {
-			shared.LogWarnf("Received value over semaphore. This should not have happened.")
-		}
-	} else {
-		lxdStorageOngoingOperationMap[imageStoragePoolLockID] = make(chan bool)
-		lxdStorageMapLock.Unlock()
+	lxdContainerCreationMap[containerCreateFromImagePoolLockID].ref++
+	lxdContainerCreationLock.Unlock()
 
-		var imgerr error
-		if !s.zfsFilesystemEntityExists(fsImage, true) {
-			imgerr = s.ImageCreate(fingerprint)
-		}
+	// ImageCreate() does not need to and cannot be placed under the
+	// container lock. Otherwise we'll deadlock.
+	fsImage := fmt.Sprintf("images/%s", fingerprint)
+	err := s.ImageCreate(fingerprint)
```

Comment:
Note: Open to race, `ImageDelete()` might grab a lock right after and delete the image.